### PR TITLE
Support redirect URI client policy executor

### LINKED ISSUE #14381: Support redirect URI client policy executor
### Description

See the entire background from #14380

### Redirect URI Client Policy Executor

Implement a configurable Redirect URI client policy that allows configuring the policy around redirect-uris, maybe options like:

Option | Type | Default | Notes
-- | -- | -- | --
Allow Private-Use URI Scheme Redirection | boolean | false | https://www.rfc-editor.org/rfc/rfc8252#section-7.1
Allow Loopback Interface Redirection | boolean | true | https://www.rfc-editor.org/rfc/rfc8252#section-7.3
Allow HTTP | boolean | false |  
Allow wildcard in context-path | boolean | true | Mainly for compatibility with Keycloak adapters
Permitted domains | List |   | List permitted domains, with support for regex, if enabled will allow wildcard in hostname, and check that configured redirect-uri matches one of the patterns
Allow Open Redirect | boolean | false |  

The client policy would only check the configured redirect-uris for a client when created or updated.

### Design

1. implement `ClientPolicyExecutorProvider` interface for **`RedirectUriClientPolicyExecutor`**
2. execute on event
    1. `REGISTER`
    2. `UPDATE`
3. with configuration options from above

### PR DESCRIPTION
**Pull request by lexcao**
Closes #14381

### CODE REVIEW

**Review by 1-alex98 — APPROVED** [review]
Even though this Pull Request does not yet solve the issue to enforce the defaults on all clients. It is well done and ready to merge/review.

**Comment by 1-alex98** [comment]
@mposolda @stianst Could you have a look at this? I can also replay this PR with conflicts resolved and commit signed but there is little benefit in it if that means the new PR will lay arround 1 and a half years as well.

File: services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureRedirectUrisExecutor.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    void validate(String redirectUri) throws ClientPolicyException {
+        UriValidation validation;
+        try {
+            validation = new UriValidation(redirectUri);
+        } catch (URISyntaxException e) {
+            throw invalidRedirectUri("Invalid Redirect Uri: invalid uri syntax");
+        }
+
+        if (configuration.allowOpen) {
+            if (!"dev".equals(System.getProperty("kc.profile"))) {
```
Comment (1-alex98) [inline services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureRedirectUrisExecutor.java:50]:
```suggestion
            if(!Environment.isDevProfile()){
```

Comment (lexcao) [reply services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureRedirectUrisExecutor.java:50]: Hi @1-alex98 
Thanks, currently `Environment.isDevProfile` is from `keycloak-quarkus-server` module which is not included in the `keycloak-services` module.

I think it better keep as it is and wait for Keycloak team to check.

**Comment by mposolda** [comment]
@lexcao Thanks and sorry for late review. There was other PR (which was based on your PR): https://github.com/keycloak/keycloak/issues/25637 . That one was already merged in Keycloak main and will be available in Keycloak 24. So I am closing this PR.

Unfortunately, your name disappeared from being co-author of the original commit, but I will make sure that you are mentioned in the release notes regarding this contribution.

**Pull request closed without merging**